File: src/main/java/com/github/wolray/line/io/LineReader.java
package com.github.wolray.line.io;

import com.alibaba.fastjson.JSON;
import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.usermodel.Workbook;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import java.io.*;
import java.util.Iterator;
import java.util.function.Function;
import java.util.function.Predicate;
import java.util.stream.Stream;

/**
 * @author wolray
 */
public class LineReader<S, V, T> {
    protected final Function<V, T> function;

    protected LineReader(Function<V, T> function) {
        this.function = function;
    }

    public static <T> Text<T> simple(Function<String, T> parser) {
        return new Text<>(parser);
    }

    public static <T> Text<T> byJson(Class<T> type) {
        return new Text<>(s -> JSON.parseObject(s, type));
    }

    public static <T> CsvReader<T> byCsv(String sep, Class<T> type) {
        return new CsvReader<>(new ValuesConverter.Text<>(new TypeValues<>(type)), sep);
    }

    public static <T> Excel<T> byExcel(Class<T> type) {
        return byExcel(0, type);
    }

    public static <T> Excel<T> byExcel(int sheetIndex, Class<T> type) {
        return new Excel<>(sheetIndex, new ValuesConverter.Excel<>(new TypeValues<>(type)));
    }

    public static InputStream toInputStream(String file) {
        try {
            return new FileInputStream(file);
        } catch (FileNotFoundException e) {
            return null;
        }
    }

    public Session read(S source) {
        return read(source, 0);
    }

    public Session read(S source, int skipLines) {
        return new Session(source, skipLines);
    }

    protected Iterator<V> toIterator(S source) {
        throw new UnsupportedOperationException();
    }

    protected void reorder(int[] slots) {
        throw new UnsupportedOperationException();
    }

    public interface SizedIterator<T> extends Iterator<T> {
        long size();
    }

    public static class Text<T> extends LineReader<InputStream, String, T> {
        Text(Function<String, T> parser) {
            super(parser);
        }

        @Override
        protected Iterator<String> toIterator(InputStream source) {
            return IteratorHelper.toIterator(new BufferedReader(new InputStreamReader(source)));
        }
    }

    public static class Excel<T> extends LineReader<InputStream, Row, T> {
        private final int sheetIndex;

        private Excel(int sheetIndex, ValuesConverter.Excel<T> converter) {
            super(converter);
            this.sheetIndex = sheetIndex;
        }

        @Override
        protected Iterator<Row> toIterator(InputStream source) {
            try {
                Workbook workbook = new XSSFWorkbook(source);
                Sheet sheet = workbook.getSheetAt(sheetIndex);
                return sheet.iterator();
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        }

        @Override
        protected void reorder(int[] slots) {
            ((ValuesConverter.Excel<T>)function).resetOrder(slots);
        }
    }

    public class Session {
        private final S source;
        private final int skipLines;
        private Predicate<V> predicate;
        private int[] slots;

        protected Session(S source, int skipLines) {
            this.source = source;
            this.skipLines = skipLines;
        }

        public Session columns(int... slots) {
            this.slots = slots;
            return this;
        }

        public Session columns(String excelCols) {
            if (excelCols == null || excelCols.isEmpty()) {
                slots = new int[0];
            } else {
                String[] split = excelCols.split(",");
                slots = new int[split.length];
                char a = 'A';
                for (int i = 0; i < split.length; i++) {
                    String col = split[i].trim();
                    int j = col.charAt(0) - a;
                    if (col.length() > 1) {
                        slots[i] = (j + 1) * 26 + col.charAt(1) - a;
                    } else {
                        slots[i] = j;
                    }
                }
            }
            return this;
        }

        protected void preprocess(Iterator<V> iterator) {
            if (slots != null && slots.length > 0) {
                reorder(slots);
            }
        }

        public Session filter(Predicate<V> predicate) {
            return filterIf(true, predicate);
        }

        public Session filterIf(boolean condition, Predicate<V> predicate) {
            if (condition) {
                this.predicate = predicate;
            }
            return this;
        }

        public DataStream<T> stream() {
            return DataStream.of(() -> {
                Iterator<V> iterator = toIterator(source);
                for (int i = 0; i < skipLines; i++) {
                    iterator.next();
                }
                preprocess(iterator);
                Long size = null;
                if (iterator instanceof SizedIterator) {
                    size = ((SizedIterator<V>)iterator).size();
                }
                Stream<V> stream = IteratorHelper.toStream(iterator, size);
                if (predicate != null) {
                    stream = stream.filter(predicate);
                }
                return stream.map(function);
            });
        }
    }
}

File: src/main/java/com/github/wolray/line/io/LineReader.kt
package com.github.wolray.line.io

import com.alibaba.fastjson.JSON
import com.github.wolray.line.io.TextScope.toSequence
import org.apache.poi.ss.usermodel.Row
import org.apache.poi.xssf.usermodel.XSSFWorkbook
import java.io.InputStream
import java.util.*
import java.util.function.Function
import java.util.function.Supplier
import kotlin.streams.asStream

/**
 * @author wolray
 */
abstract class LineReader<S, V, T>(val function: Function<V, T>) : IReader<S, V, T> {

    override fun read(source: S) = Session(source)
    override fun invoke(v: V): T = function.apply(v)

    open inner class Session(private val source: S) : Chainable<Session> {
        override val self get() = this
        private var errorType: Class<out Throwable>? = null
        private var skip: Int = 0
        protected var slots: IntArray? = null
        protected var cols: Array<String>? = null

        fun ignoreError(type: Class<out Throwable>) = apply { errorType = type }
        fun skipLines(n: Int) = apply { skip = n }

        fun columns(vararg col: String) = apply { cols = arrayOf(*col) }
        fun columns(vararg index: Int) = apply { slots = index }
        fun columnsBefore(index: Int) = columnsRange(0, index)

        fun columnsRange(start: Int, before: Int) = apply {
            slots = (start until before).toList().toIntArray()
        }

        fun excelColumns(excelCols: String) = apply {
            slots = if (excelCols.isEmpty()) intArrayOf() else {
                excelCols.split(",").map(::excelIndex).toIntArray()
            }
        }

        private fun excelIndex(col: String): Int {
            return col.trim().fold(0) { acc, c -> acc * 26 + (c - A + 1) } - 1
        }

        protected open fun preprocess(iterator: Iterator<V>) {}

        private fun getIterator(): Iterator<V> {
            return try {
                toIterator(source).apply {
                    if (skip > 0) repeat(skip) { next() }
                    preprocess(this)
                }
            } catch (e: Throwable) {
                if (errorType?.isAssignableFrom(e.javaClass) != true) throw e
                else Collections.emptyIterator()
            }
        }

        fun sequence(): Sequence<T> {
            return Sequence(::getIterator).map(this@LineReader)
        }

        fun <R> sequence(function: Function<Sequence<T>, R>): R {
            return function.apply(sequence())
        }

        fun stream(): DataStream<T> = DataStream.of { sequence().asStream() }
    }

    open class Simple<T>(function: Function<String, T>) :
        LineReader<Text, String, T>(function), IReader.Is<String, T> {

        override fun toIterator(source: Text): Iterator<String> {
            return source.toSequence().iterator()
        }
    }

    class Excel<T> internal constructor(
        private val sheetIndex: Int,
        converter: ValuesConverter.Excel<T>
    ) : ValuesReader<Text, Row, T>(converter), IReader.Is<Row, T> {

        override fun toIterator(source: Supplier<InputStream>): Iterator<Row> =
            XSSFWorkbook(source.get()).getSheetAt(sheetIndex).iterator()

        override fun splitHeader(iterator: Iterator<Row>) = iterator.next().map { it.stringCellValue }
    }

    companion object {
        private const val A = 'A'

        @JvmStatic
        fun <T> simple(parser: Function<String, T>) = Simple(parser)

        @JvmStatic
        fun <T> byJson(type: Class<T>) = Simple { JSON.parseObject(it, type) }

        @JvmStatic
        fun <T> byCsv(sep: String, type: Class<T>): CsvReader<T> {
            return CsvReader(ValuesConverter.Csv(TypeValues(type)), sep)
        }

        @JvmStatic
        fun <T> byExcel(type: Class<T>) = byExcel(0, type)

        @JvmStatic
        fun <T> byExcel(sheetIndex: Int, type: Class<T>): Excel<T> {
            return Excel(sheetIndex, ValuesConverter.Excel(TypeValues(type)))
        }
    }
}
